include/fmt/ranges.h

```diff
@@ -0,0 +1,226 @@
+template <typename T>
+class is_tuple_like_ {
+  template <typename U>
+  static auto check(U *p)
+      -> decltype(std::tuple_size<U>::value,
+                  std::declval<typename std::tuple_element<0, U>::type>(), int());
+  template <typename>
+  static void check(...);
+
+ public:
+  static FMT_CONSTEXPR_DECL const bool value = !std::is_void<decltype(check<T>(nullptr))>::value;
+};
+
+template <typename T>
+struct is_tuple_like {
+  static FMT_CONSTEXPR_DECL const bool value = is_tuple_like_<T>::value && !is_range_<T>::value;
+};
+
+//=--------------------------------------------------------------------------------------------------------------------
+template <size_t... Is, class Tuple, class F>
+void for_each(std::index_sequence<Is...>, Tuple &&tup, F &&f) noexcept {
+  using std::get;
+  // using free function get<I>(T) now.
+  const int _[] = {0, ((void)f(get<Is>(tup)), 0)...};
+  (void)_;  // blocks warnings
+}
+//=--------------------------------------------------------------------------------------------------------------------
+template <class T>
+FMT_CONSTEXPR std::make_index_sequence<std::tuple_size<T>::value> 
+get_indexes(T const &) { return {}; }
+
+//=--------------------------------------------------------------------------------------------------------------------
+template <class Tuple, class F>
+void for_each(Tuple &&tup, F &&f) {
+  const auto indexes = get_indexes(tup);
+  for_each(indexes, std::forward<Tuple>(tup), std::forward<F>(f));
+}
+
+}  // namespace internal
+
+// =====================================================================================================================
+template <typename TupleT, typename Char>
+struct formatter<TupleT, Char, 
+    typename std::enable_if<fmt::internal::is_tuple_like<TupleT>::value>::type> {
+
+  fmt::formatting_tuple<Char> formatting;
+
+  template <typename ParseContext>
+  FMT_CONSTEXPR auto parse(ParseContext &ctx) -> decltype(ctx.begin()) {
+    return formatting.parse(ctx);
+  }
+
+  template <typename FormatContext = format_context>
+  auto format(const TupleT &values, FormatContext &ctx) -> decltype(ctx.out()) {
+    auto out = ctx.out();
+    std::size_t i = 0;
+    internal::copy(formatting.prefix, out);
+    internal::for_each(values, [&](const auto &v) {
+      if (i++ > 0) {
+        internal::copy(formatting.delimiter, out);
+      }
+      if (formatting.add_spaces) {
+        format_to(out, " {}", v);
+      } else {
+        format_to(out, "{}", v);
+      }
+    });
+    if (formatting.add_spaces) {
+      *out++ = ' ';
+    }
+    internal::copy(formatting.postfix, out);
+
+    return ctx.out();
+  }
+};
+
+template <typename RangeT, typename Char>
+struct formatter< RangeT, Char,
+    typename std::enable_if<fmt::internal::is_range<RangeT>::value>::type> {
+
+  static FMT_CONSTEXPR_DECL const std::size_t range_length_limit =
+      FMT_RANGE_OUTPUT_LENGTH_LIMIT; // output only up to N items from the range.
```

Comment:
Why do we need a limit? User can always pass a subrange to `fmt::join` if they don't want to format everything.

Reply:
Because it could be and infinite range, or nested range where user can not use fmt::join.
`std::tuple<std::vector<std::vector<int>>>`